blockchain/utxocache.go

```diff
@@ -0,0 +1,1085 @@
+type utxoView interface {
+	// getEntry tries to get an entry from the view.  If the entry is not
+	// in the view, both the returned entry and the error are nil.
+	getEntry(outpoint wire.OutPoint) (*UtxoEntry, error)
+
+	// addEntry adds a new entry to the view.  Set overwrite to true if
+	// this entry should overwrite any existing entry for the same
+	// outpoint.
+	addEntry(outpoint wire.OutPoint, entry *UtxoEntry, overwrite bool) error
+
+	// spendEntry marks an entry as spent.
+	spendEntry(outpoint wire.OutPoint, entry *UtxoEntry) error
+}
+
+// utxoByHashSource is an interface that allows fetching UTXO entries by
+// transaction hash.
+// This interface exists due to the legacy spend journal database structure.
+type utxoByHashSource interface {
+	// getEntryByHash looks for an entry with the given transaction hash.
+	// This method exists due to the legacy spend journal database structure.
+	// Its execution is very inefficient, but it's almost never used.
+	getEntryByHash(hash *chainhash.Hash) (*UtxoEntry, error)
+}
+
+// utxoBatcher is an interface that allows a caller to batch fetch UTXOs from
+// an underlying view.
+type utxoBatcher interface {
+	utxoView
+
+	// getEntries attempts to fetch a serise of entries from the UTXO view.
+	// If a single entry is not found within the view, then an error is
+	// returned.
+	getEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error)
+}
+
+// mapSlice is a slice of maps for Utxo entries.  The slice of maps are needed to
+// guarantee that the map will only take up N amount of bytes.  As of v1.20, the
+// go runtime will allocate 2^N + few extra buckets, meaning that for large N, we'll
+// allocate a lot of extra memory if the amount of entries goes over the previously
+// allocated buckets.  A slice of maps allows us to have a better control of how much
+// total memory gets allocated by all the maps.
+type mapSlice struct {
+	maps []map[wire.OutPoint]*UtxoEntry
+
+	// maxEntries is the maximum amount of elemnts that the map is allocated for.
+	maxEntries []int
+
+	// maxTotalMemoryUsage is the maximum memory usage in bytes that the state
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+}
+
+// length returns the length of all the maps in the map slice added together.
+func (ms *mapSlice) length() int {
+	var l int
+	for _, m := range ms.maps {
+		l += len(m)
+	}
+
+	return l
+}
+
+// size returns the size of all the maps in the map slice added together.
+func (ms *mapSlice) size() int {
+	var size int
+	for _, num := range ms.maxEntries {
+		size += CalculateRoughMapSize(num, bucketSize)
+	}
+
+	return size
+}
+
+// get looks for the outpoint in all the maps in the map slice and returns
+// the entry.  nil and false is returned if the outpoint is not found.
+func (ms *mapSlice) get(op wire.OutPoint) (*UtxoEntry, bool) {
+	var entry *UtxoEntry
+	var found bool
+
+	for _, m := range ms.maps {
+		entry, found = m[op]
+		if found {
+			break
```

Comment:
style nit: return early with the found element?

Reply:
Like `return entry, found` instead of `break`?

Reply:
yep